api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -98,12 +101,62 @@ public Workspace apply(org.pmiops.workbench.db.model.Workspace workspace) {
           }
 
-          
+
           result.setUserRoles(workspace.getWorkspaceUserRoles().stream().map(TO_CLIENT_USER_ROLE).collect(Collectors.toList()));
 
           return result;
         }
       };
+      private static final BiFunction<org.pmiops.workbench.db.model.Workspace, org.pmiops.workbench.firecloud.model.Workspace, Workspace>
+          TO_CLIENT_WORKSPACE_FROM_FC_AND_DB =
+          new BiFunction<org.pmiops.workbench.db.model.Workspace, org.pmiops.workbench.firecloud.model.Workspace, Workspace>() {
+            @Override
+            public Workspace apply(org.pmiops.workbench.db.model.Workspace workspace,
+                org.pmiops.workbench.firecloud.model.Workspace fcWorkspace) {
+              ResearchPurpose researchPurpose = new ResearchPurpose()
+                  .diseaseFocusedResearch(workspace.getDiseaseFocusedResearch())
+                  .diseaseOfFocus(workspace.getDiseaseOfFocus())
+                  .methodsDevelopment(workspace.getMethodsDevelopment())
+                  .controlSet(workspace.getControlSet())
+                  .aggregateAnalysis(workspace.getAggregateAnalysis())
+                  .ancestry(workspace.getAncestry())
+                  .commercialPurpose(workspace.getCommercialPurpose())
+                  .population(workspace.getPopulation())
+                  .populationOfFocus(workspace.getPopulationOfFocus())
+                  .additionalNotes(workspace.getAdditionalNotes())
+                  .reviewRequested(workspace.getReviewRequested())
+                  .approved(workspace.getApproved());
+
+              if(workspace.getTimeRequested() != null){
+                researchPurpose.timeRequested(workspace.getTimeRequested().getTime());
+              }
+
+              if(workspace.getTimeReviewed() != null){
+                researchPurpose.timeReviewed(workspace.getTimeReviewed().getTime());
+              }
+              Workspace result = new Workspace()
+                  .etag(Etags.fromVersion(workspace.getVersion()))
+                  .lastModifiedTime(workspace.getLastModifiedTime().getTime())
+                  .creationTime(workspace.getCreationTime().getTime())
+                  .dataAccessLevel(workspace.getDataAccessLevel())
+                  .name(workspace.getName())
+                  .id(fcWorkspace.getName())
+                  .namespace(fcWorkspace.getNamespace())
+                  .description(workspace.getDescription())
+                  .researchPurpose(researchPurpose);
+              if (fcWorkspace.getCreatedBy() != null) {
+                result.setCreator(fcWorkspace.getCreatedBy());
+              }
+              if (workspace.getCdrVersion() != null) {
+                result.setCdrVersionId(String.valueOf(workspace.getCdrVersion().getCdrVersionId()));
+              }
+
+
+              result.setUserRoles(workspace.getWorkspaceUserRoles().stream().map(TO_CLIENT_USER_ROLE).collect(Collectors.toList()));
```

Comment:
Should we use fcWorkspace for the roles?

Reply:
You don't actually get the roles from getWorkspace in FC, just your own role. I am planning on using that for enforcing write access or higher.

Reply:
SGTM. Thanks!